src/stream/src/executor_v2/lookup/impl_.rs

```diff
@@ -195,21 +200,38 @@ impl<S: StateStore> LookupExecutor<S> {
     /// messages until there's one.
     ///
     /// If we can use `async_stream` to write this part, things could be easier.
-    pub async fn next_inner(&mut self) -> Result<Option<Message>> {
-        match self.input.next().await.expect("unexpected end of stream")? {
-            ArrangeMessage::Barrier(barrier) => {
-                self.process_barrier(barrier.clone()).await?;
-                Ok(Some(Message::Barrier(barrier)))
+    #[try_stream(ok = Message, error = TracedStreamExecutorError)]
+    pub async fn next_execute(mut self: Box<Self>) {
+        loop {
+            let msg = self
+                .input
+                .next()
+                .await
+                .expect("unexpected end of input")
+                .map_err(StreamExecutorError::input_error)?;
```

Comment:
Please use `#[for_await]` here.

Reply:
When I try to use for_await, the compiler complains about partial move. Any idea on how to solve it?
```
error[E0382]: borrow of partially moved value: `*self`
   --> src/stream/src/executor_v2/lookup/impl_.rs:216:21
    |
206 |         for msg in self.input {
    |                    ---------- value partially moved here
...
216 |                     self.process_barrier(barrier.clone())
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ value borrowed here after partial move
    |
    = note: partial move occurs because `self.input` has type `std::pin::Pin<std::boxed::Box<dyn futures::Stream<Item = std::result::Result<executor_v2::lookup::sides::ArrangeMessage, risingwave_common::error::RwError>> + std::marker::Send>>`, which does not implement the `Copy` trait
```
I think a workaround is to add a wrapper to contain the lookup executor and its `input`, thus moving `input` and borrowing executor will not cause above problem, like what `executor_v2/agg.rs` does.

Reply:
Try `for msg in &mut self.input`?

Reply:
When using `&mut self.input`, it complains about borrowing as mut more than once.
```
error[E0499]: cannot borrow `*self` as mutable more than once at a time
   --> src/stream/src/executor_v2/lookup/impl_.rs:210:21
    |
71  | |     pub arrangement_order_rules: Vec<OrderPair>,
    | |_______________________________________________________________________- first borrow later used here
...
203 | /     #[try_stream(ok = Message, error = TracedStreamExecutorError)]
...
206 |           for msg in &mut self.input {
    |                      --------------- first mutable borrow occurs here
...
210 |                       self.process_barrier(barrier.clone())
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ second mutable borrow occurs here
```

I thought of an alternative workaround. I can move all the fields to local variables and extract all the methods to functions to avoid borrowing `self` when generating stream messages.

Reply:
That's okay. On the other side, we may make `self.input` an `Option` and take it first.